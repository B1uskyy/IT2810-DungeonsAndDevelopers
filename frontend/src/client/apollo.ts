import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
