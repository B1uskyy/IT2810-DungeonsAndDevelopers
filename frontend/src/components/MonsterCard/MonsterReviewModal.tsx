import { useMutation, useQuery } from '@apollo/client';
import { Box, Button, Dialog, DialogActions, DialogContent, DialogContentText } from '@mui/material';
import React, { useContext, useEffect, useState } from 'react';
import { AuthContext } from '../../context/AuthContext';
import { useToast } from '../../hooks/useToast';
import { ReviewFormType, ReviewType } from '../../interfaces/ReviewProps.ts';
import ReviewSlider from './ReviewSlider.tsx';
import ReviewTextField from './ReviewTextField.tsx';
import { GET_MONSTER_REVIEWS } from '../../graphql/getMonsterQuerie.ts';
import { ADD_REVIEW, UPDATE_REVIEW } from '../../graphql/reviewQueries.ts';

const MonsterReviewModal = ({ name, monsterId, image }: ReviewFormType) => {
  const { userId } = useContext(AuthContext);
  const [isOpen, setIsOpen] = useState(false);
  const [editMode, setEditMode] = useState(false);
  const [difficulty, setDifficulty] = useState(50);
  const [description, setDescription] = useState('');
  const [originalDifficulty, setOriginalDifficulty] = useState(50);
  const [originalDescription, setOriginalDescription] = useState('');
  const { showToast } = useToast();

  const { data } = useQuery(GET_MONSTER_REVIEWS, { variables: { monsterId } });
  const existingReview = data?.monster?.reviews.find((review: ReviewType) => review.user.id === userId);

  const [addReview] = useMutation(ADD_REVIEW, {
    refetchQueries: [{ query: GET_MONSTER_REVIEWS, variables: { monsterId } }],
  });

  const [updateReview] = useMutation(UPDATE_REVIEW, {
    refetchQueries: [{ query: GET_MONSTER_REVIEWS, variables: { monsterId } }],
  });

  useEffect(() => {
    if (existingReview) {
      setDescription(existingReview.description);
      setDifficulty(existingReview.difficulty);
      setOriginalDescription(existingReview.description);
      setOriginalDifficulty(existingReview.difficulty);
    }
  }, [existingReview]);

  const handleClickOpen = () => {
    if (existingReview) {
      setEditMode(true);
    } else {
      setDescription('');
      setDifficulty(50);
      setEditMode(false);
    }
    setIsOpen(true);
  };

  const handleClose = () => {
    if (editMode) {
      setDescription(originalDescription);
      setDifficulty(originalDifficulty);
    } else {
      setDescription('');
      setDifficulty(50);
    }
    setIsOpen(false);
    setEditMode(false);
  };

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();
    if (!userId) {
      showToast({ message: 'You must be logged in to submit a review', type: 'error', duration: 3000 });
      return;
    }

    try {
      if (editMode && existingReview) {
        // Update existing review
        await updateReview({
          variables: {
            monsterId,
            reviewId: existingReview.id,
            review: { user: userId, difficulty, description },
          },
        });
        showToast({ message: `Review on ${name} updated`, type: 'success', duration: 3000 });
      } else {
        // Add new review
        await addReview({
          variables: {
            monsterId,
            review: { user: userId, difficulty, description },
          },
        });
        showToast({ message: `Review on ${name} submitted`, type: 'success', duration: 3000 });
      }
      setIsOpen(false);
    } catch (error) {
      console.error('Error submitting review:', error);
      showToast({ message: `Failed to submit review on ${name}`, type: 'error', duration: 3000 });
    }
  };

  return (
    <>
      <Button
        variant="outlined"
        onClick={handleClickOpen}
        sx={{
          color: 'white',
          borderColor: '#DB3232',
          backgroundColor: '#DB3232',
          padding: '0.2vw 0.5vw',
          minWidth: 'unset',
          minHeight: 'unset',
          height: 'auto',
          lineHeight: 'normal',
          borderRadius: '4px',
          boxSizing: 'border-box',
          '&:hover': {
            borderColor: '#DB3232',
            backgroundColor: 'black',
            color: '#DB3232',
          },
          fontFamily: 'MedievalSharp',
          fontSize: {
            xs: '3.5vw',
            sm: '2vw',
            md: '1.3vw',
            lg: '1.2vw',
            xl: '1vw',
          },
          textTransform: 'none',
        }}
        aria-label="Review button"
      >
        Review
      </Button>
      <Dialog
        open={isOpen}
        onClose={handleClose}
        PaperProps={{
          component: 'form',
          sx: {
            width: { xs: '100vw', sm: '95vw', md: '90vw', lg: '80vw', xl: '70vw' },
            height: { xs: '100vh', sm: '85vh', md: '85vh', lg: '60vh', xl: '70vh' },
            maxWidth: 'none',
            padding: { xs: 0, sm: 4 },
            backgroundColor: 'black',
          },
        }}
      >
        <DialogContent className="flex flex-col xl:flex-row xl:items-center bg-black gap-6">
          <Box className="w-full xl:w-1/2 flex flex-col gap-[50px] xl:items-center">
            <img src={image} alt="Image of selected monster" className="w-full sm:w-3/4 xl:w-1/2 rounded" />
            <h2 className="sub-header text-white">Review of {name}</h2>
          </Box>
          <article className="flex flex-col gap-4 w-full xl:w-1/2">
            <DialogContentText sx={{ color: 'white', fontSize: '1.5rem', fontFamily: 'MedievalSharp' }}>
              Difficulty
            </DialogContentText>
            <ReviewSlider value={difficulty} onChange={(_, value) => setDifficulty(value as number)} />
            <ReviewTextField value={description} onChange={(e) => setDescription(e.target.value.slice(0, 300))} />
          </article>
        </DialogContent>

        <DialogActions className="bg-black">
          <Button
            onClick={handleClose}
            aria-label="Cancel-button"
            sx={{
              color: 'white',
              borderColor: 'white',
              '&:hover': {
                borderColor: '#DB3232',
                color: '#DB3232',
              },
              fontFamily: 'MedievalSharp',
              fontSize: '1.5rem',
            }}
          >
            Cancel
          </Button>
          <Button
            onClick={handleSubmit}
            type="submit"
            aria-label="Save-button"
            sx={{
              color: 'white',
              borderColor: 'white',
              '&:hover': {
                borderColor: '#DB3232',
                color: '#DB3232',
              },
              fontFamily: 'MedievalSharp',
              fontSize: '1.5rem',
            }}
          >
            {editMode ? 'UPDATE' : 'SUBMIT'}
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default MonsterReviewModal;
