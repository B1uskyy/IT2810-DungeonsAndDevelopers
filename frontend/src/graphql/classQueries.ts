import { gql } from '@apollo/client';

export const GET_CLASSES = gql`
  query GetClasses($offset: Int, $limit: Int) {
    classes(offset: $offset, limit: $limit) {
      classes {
        id
        index
        name
        hit_die
        skills
      }
      totalClasses
    }
  }
`;
