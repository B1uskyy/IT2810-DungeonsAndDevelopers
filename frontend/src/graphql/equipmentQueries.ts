import { gql } from '@apollo/client';

export const ADD_EQUIPMENT_TO_CHARACTER = gql`
  mutation AddEquipmentToCharacter($userId: ID!, $equipmentId: ID!) {
    addEquipmentToCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const REMOVE_EQUIPMENT_FROM_CHARACTER = gql`
  mutation RemoveEquipmentFromCharacter($userId: ID!, $equipmentId: ID!) {
    removeEquipmentFromCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const GET_EQUIPMENT_SUGGESTIONS = gql`
  query GetEquipmentSuggestions($searchTerm: String!) {
    equipments(searchTerm: $searchTerm, suggestionsOnly: true) {
      equipments {
        id
        name
      }
      totalCount
    }
  }
`;

export const REMOVE_ALL_EQUIPMENTS = gql`
  mutation RemoveAllEquipments($userId: ID!) {
    removeAllEquipments(userId: $userId) {
      id
      equipments {
        id
        name
        category
        value
      }
    }
  }
`;
export const GET_EQUIPMENTS = gql`
  query GetEquipments($searchTerm: String, $offset: Int, $limit: Int) {
    equipments(searchTerm: $searchTerm, offset: $offset, limit: $limit) {
      equipments {
        id
        index
        name
        category
        value
      }
      totalCount
    }
  }
`;
