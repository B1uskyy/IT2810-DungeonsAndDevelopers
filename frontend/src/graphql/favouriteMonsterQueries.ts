import { gql } from '@apollo/client';

export const ADD_FAVORITE_MONSTER = gql`
  mutation AddFavoriteMonster($userId: ID!, $monsterId: ID!) {
    addFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;

export const REMOVE_FAVORITE_MONSTER = gql`
  mutation RemoveFavoriteMonster($userId: ID!, $monsterId: ID!) {
    removeFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;
