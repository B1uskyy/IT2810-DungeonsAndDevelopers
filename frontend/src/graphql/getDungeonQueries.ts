import { gql } from '@apollo/client';

export const GET_USER_DUNGEON = gql`
  query GetUserDungeon($userId: ID!) {
    user(id: $userId) {
      dungeonName
      favoritedMonsters {
        id
      }
    }
  }
`;
