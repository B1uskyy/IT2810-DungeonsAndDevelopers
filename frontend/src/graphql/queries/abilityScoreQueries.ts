import { gql } from '@apollo/client';

export const GET_ABILITYSCORES = gql`
  query GetAbilityScores($offset: Int, $limit: Int) {
    abilities(offset: $offset, limit: $limit) {
      abilities {
        index
        name
        skills
      }
      totalAbilities
    }
  }
`;
