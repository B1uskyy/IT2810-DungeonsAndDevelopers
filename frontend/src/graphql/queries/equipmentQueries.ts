import { gql } from '@apollo/client';

export const GET_EQUIPMENT_SUGGESTIONS = gql`
  query GetEquipmentSuggestions($searchTerm: String!) {
    equipments(searchTerm: $searchTerm, suggestionsOnly: true) {
      equipments {
        id
        name
      }
      totalCount
    }
  }
`;

export const GET_EQUIPMENTS = gql`
  query GetEquipments($searchTerm: String, $offset: Int, $limit: Int) {
    equipments(searchTerm: $searchTerm, offset: $offset, limit: $limit) {
      equipments {
        id
        index
        name
        category
        value
      }
      totalCount
    }
  }
`;
