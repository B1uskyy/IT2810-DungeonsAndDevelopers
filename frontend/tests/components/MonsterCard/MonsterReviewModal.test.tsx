import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MockedProvider } from '@apollo/client/testing';
import MonsterReviewModal from '../../../src/components/MonsterCard/MonsterReviewModal.tsx';
import { ADD_REVIEW, GET_MONSTER_REVIEWS, UPDATE_REVIEW } from '../../../../backend/src/graphql/queries';
import { AuthContext } from '../../../src/context/AuthContext.tsx';

// Mock the useToast hook
vi.mock('../../../src/hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: vi.fn(),
  }),
}));

const mockProps = {
  name: 'Dragon',
  monsterId: '123',
  image: 'dragon.jpg',
};

const mockUserId = 'user123';
const mockToken = 'mock-token';
const mockUsername = 'bob';
const mockLogin = vi.fn();
const mockLogout = vi.fn();

const mockReviews = [
  {
    id: 'review1',
    user: { id: mockUserId, userName: mockUsername },
    difficulty: 75,
    description: 'Existing review',
    createdAt: '2024-11-05T12:00:00Z',
  },
];

// Apollo query mocks
const mocks = [
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          reviews: mockReviews,
        },
      },
    },
  },
  {
    request: {
      query: ADD_REVIEW,
      variables: {
        monsterId: mockProps.monsterId,
        review: {
          user: mockUserId,
          difficulty: 50,
          description: 'New review',
        },
      },
    },
    result: {
      data: {
        addReview: {
          reviews: [
            {
              user: {
                id: mockUserId,
                userName: mockUsername,
              },
              difficulty: 50,
              description: 'New review',
              createdAt: '2024-11-06T08:00:00Z',
            },
          ],
        },
      },
    },
  },
  {
    request: {
      query: UPDATE_REVIEW,
      variables: {
        monsterId: mockProps.monsterId,
        reviewId: 'review1',
        review: {
          user: mockUserId,
          difficulty: 75,
          description: 'Updated review',
        },
      },
    },
    result: {
      data: {
        updateReview: {
          id: 'review1',
          user: {
            id: mockUserId,
            userName: mockUsername,
          },
          difficulty: 75,
          description: 'Updated review',
          createdAt: '2024-11-06T08:00:00Z',
        },
      },
    },
  },
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          reviews: mockReviews,
        },
      },
    },
  },
];

const renderComponent = () => {
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AuthContext.Provider
        value={{ userId: mockUserId, userName: mockUsername, logout: mockLogout, login: mockLogin, token: mockToken }}
      >
        <MonsterReviewModal {...mockProps} />
      </AuthContext.Provider>
    </MockedProvider>
  );
};

describe('MonsterReviewModal', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders review button correctly', () => {
    const { container } = renderComponent();
    expect(screen.getByText('Review')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('opens modal when review button is clicked', async () => {
    renderComponent();
    const user = userEvent.setup();

    await user.click(screen.getByText('Review'));

    expect(screen.getByText(`Review of ${mockProps.name}`)).toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute('src', mockProps.image);
    expect(screen.getByText('Difficulty')).toBeInTheDocument();
  });

  it('shows existing review data for user who has already reviewed', async () => {
    renderComponent();
    const user = userEvent.setup();

    await user.click(screen.getByText('Review'));

    await waitFor(() => {
      expect(screen.getByRole('textbox')).toHaveValue('Existing review');
    });
  });

  it('handles submission of new review', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider
          value={{ userId: mockUserId, token: 'user124', login: mockLogin, logout: mockLogout, userName: mockUsername }}
        >
          <MonsterReviewModal {...mockProps} />
        </AuthContext.Provider>
      </MockedProvider>
    );

    const user = userEvent.setup();

    await user.click(screen.getByText('Review'));
    await user.clear(screen.getByRole('textbox'));
    await user.type(screen.getByRole('textbox'), 'New review');
    await user.click(screen.getByRole('button', { name: 'Save-button' }));

    expect(container).toMatchSnapshot();
  });

  it('handles updating existing review', async () => {
    renderComponent();
    const user = userEvent.setup();

    await user.click(screen.getByText('Review'));
    await waitFor(() => {
      expect(screen.getByRole('textbox')).toHaveValue('Existing review');
    });

    await user.clear(screen.getByRole('textbox'));
    await user.type(screen.getByRole('textbox'), 'Updated review');
    await user.click(screen.getByRole('button', { name: 'Save-button' }));

    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  it('closes modal when cancel is clicked', async () => {
    renderComponent();
    const user = userEvent.setup();

    await user.click(screen.getByText('Review'));
    await user.click(screen.getByText('Cancel'));

    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  it(
    'handles character limit in description',
    async () => {
      renderComponent();
      const user = userEvent.setup();

      await user.click(screen.getByText('Review'));

      const longText = 'a'.repeat(301);
      await user.clear(screen.getByRole('textbox'));
      await user.type(screen.getByRole('textbox'), longText);

      expect(screen.getByRole('textbox')).toHaveValue('a'.repeat(300));
    },
    { timeout: 10000 }
  );

  it('resets form when modal is closed', async () => {
    renderComponent();
    const user = userEvent.setup();

    await user.click(screen.getByText('Review'));
    await user.clear(screen.getByRole('textbox'));
    await user.type(screen.getByRole('textbox'), 'Test review');
    await user.click(screen.getByText('Cancel'));

    await user.click(screen.getByText('Review'));

    await waitFor(() => {
      expect(screen.getByRole('textbox')).toHaveValue('Existing review');
    });
  });
});
