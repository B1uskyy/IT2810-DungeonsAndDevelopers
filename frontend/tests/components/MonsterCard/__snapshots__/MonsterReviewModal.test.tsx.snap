// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MonsterReviewModal > handles submission of new review 1`] = `
<div
  aria-hidden="true"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1l3bbxn-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Review
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    >
      <span
        class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
        style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
      >
        <span
          class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
        />
      </span>
    </span>
  </button>
</div>
`;

exports[`MonsterReviewModal > renders review button correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1l3bbxn-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Review
  </button>
</div>
`;
